vhdx/src/vhdx_header.rs

```diff
@@ -0,0 +1,536 @@
+
+use byteorder::{ByteOrder, LittleEndian, ReadBytesExt, WriteBytesExt};
+use std::collections::btree_map::BTreeMap;
+use std::fmt::{self, Display};
+use std::fs::File;
+use std::io::{self, Read, Seek, SeekFrom, Write};
+use std::mem::size_of;
+use uuid::Uuid;
+
+pub enum Error {
+    InvalidVHDXSign,
+    ReadingFileTypeIndetifier(io::Error),
+    InvalidHeaderSign,
+    ReadingHeader(io::Error),
+    InvalidRegionSign,
+    ReadingRegionTableHeader(io::Error),
+    ReadingRegionTableEntry(io::Error),
+    ReadingChecksum,
+    InvalidChecksum(String),
+    InvalidEntryIndex,
+    InvalidEntryCount,
+    InvalidUuid,
+    WritingData(io::Error),
+    WritingHeader(io::Error),
+    NoValidHeader,
+    RegionOverlap,
+    DuplicateBATEntry,
+    DuplicateMDREntry,
+    UnrecognizedRegionEntry,
+    RegionEntryCollectionFailed,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Display for Error {
+    #[remain::check]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        #[sorted]
+        match self {
+            DuplicateBATEntry => write!(f, "BAT entry is not unique"),
+            DuplicateMDREntry => write!(f, "Metadata region entry is not unique"),
+            InvalidChecksum(section) => write!(f, "{} checksum doesn't match", section),
+            InvalidEntryCount => write!(f, "invalid entry cound"),
+            InvalidEntryIndex => write!(f, "entry index is too big"),
+            InvalidHeaderSign => write!(f, "not a valid VHDX header"),
+            InvalidRegionSign => write!(f, "not a valid VHDX region"),
+            InvalidUuid => write!(f, "couldn't parse Uuid for region entry"),
+            InvalidVHDXSign => write!(f, "not a VHDX file"),
+            NoValidHeader => write!(f, "No valid header found"),
+            ReadingChecksum => write!(f, "cannot read checksum"),
+            ReadingFileTypeIndetifier(e) => write!(f, "failed to read File Type Identifier: {}", e),
+            ReadingHeader(e) => write!(f, "failed to read headers: {}", e),
+            ReadingRegionTableEntry(e) => write!(f, "failed to read region table entry: {}", e),
+            ReadingRegionTableHeader(e) => write!(f, "failed to read region table header: {}", e),
+            RegionEntryCollectionFailed => write!(f, "failed to read region entries"),
+            //UnsupportedVersion(v) => write!(f, "unsupported version: {}", v),
+            RegionOverlap => write!(f, "overlapping regions found"),
+            UnrecognizedRegionEntry => write!(f, "we do not recongize this entry"),
+            WritingData(e) => write!(f, "failed to write data: {}", e),
+            WritingHeader(e) => write!(f, "failed to write header: {}", e),
+        }
+    }
+}
+
+#[derive(Clone)]
+struct FileTypeIdentifier {
+    pub signature: u64,
+    pub creator: [u8; 512], //should we keep it here?
+}
+
+const VHDX_SIGN: u64 = 0x656C_6966_7864_6876; //"vhdxfile"
+const FILE_START: u64 = 0; //The first element
```

Comment:
Please declare any constant after the crates imports, this way they're located in a single place for the whole file.

Reply:
Done.

Reply:
This one is not resolved.